proposed/2022/dotnet-nuget-config-spec.md

```diff
@@ -0,0 +1,146 @@
+# NuGet configuration CLI for dotnet.exe
+
+- Author: [Heng Liu](https://github.com/heng-liu)
+- GitHub Issue [8420](https://github.com/NuGet/Home/issues/8420)
+
+## Problem Background
+
+Currently, there is no NuGet configuration CLI for dotnet.exe. It's inconvenient for NuGet users to know about the NuGet configuration file locations and figure out where is the merged configuration coming from. 
+
+## Who are the customers
+
+This feature is for dotnet.exe users.
+
+## Goals
+Design and implement `dotnet nuget config` command.
+
+## Non-Goals
+Design and implement `dotnet nuget config` command with commands other than `list`, e.g. add/update/delete
+
+## Solution
+The following command will be implemented in the `dotnet.exe` CLI.
+
+### `dotnet nuget config`
+
+#### Commands
+
+- List
+
+Lists all the NuGet configuration file locations. 
+
+This command will include all the NuGet configuration file that will be applied, when invoking NuGet command from the current working directory path. The listed NuGet configuration files are in priority order. So the order of loading those configurations is reversed, that is, loading order is from the bottom to the top. So the configuration on the top will apply.
+You may refer to [How settings are applied](https://learn.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior#how-settings-are-applied) for more details. 
+
+#### Arguments
+
+- WORKING_DIRECTORY
+
+Run this command as if working directory is set to the specified directory.
+
+If the specified `WORKING_DIRECTORY` doesn't exist, an error is displayed indicating the `WORKING_DIRECTORY` doesn't exist.
+
+> [!Note]
+> If `WORKING_DIRECTORY` (or its parent directories) is not accessible, the command will ignore any NuGet configuration files under those directories without any warning/error. This is aligned with other NuGet commands.
+
+#### Options
+- -v|--verbosity <LEVEL>
+
+Sets the verbosity level of the command. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic]. The default is minimal. 
+When the verbosity level is detailed or diagnostic, the command will display the merged NuGet configuration.
```

Comment:
It feels out of place to me to implement "display merged config" as the detailed output for "Lists all the NuGet configuration file locations". 

I feel like the functionality belongs in a `get` verb, not `list`.

I have no idea if anyone will actually want to do it, but also imagine that someone wants to have a script where the output is piped to a file, and then they use an XML processing library/tool to work on the XML.  With this design, it's difficult because of all the non-XML output. Whereas a dedicated command to get the merged config could have only XML output at minimal verbosity, allowing customers to "safely" process the XML output of something like `dotnet nuget config get all -v m`.

You might also want to have a look at `git config` and all its options as prior-art that customers might already be familiar with. Admittedly they use `--list` to list all the config values (what you're calling the "merged config"), rather than listing all the files that contributed. I don't see an option to only list gitconfig files used, but they do have a `--show-origin`, so `git config --list --show-origin` will show all the config key-value pairs, and which

Reply:
I agree, verbosity should only really be a thing when you want to know more about certain config value. 

IMO here's how it should work longer term: 

- dotnet nuget list config-> Lists all NuGet config file paths.
- dotnet nuget list config globalPackagesFolder -> Gives you the value of the globalPackagesFolder configuration. 
- dotnet nuget add config globalPackagesFolder D:\packages -> Sets the global packages folder in the approrpiate config. (same thing would apply to update)
- dotnet nuget list config globalPackagesFolder --verbosity detailed -> Gives you the value of the global packages folder configuration + the config where it comes from.

I know nuget.exe uses `get` and `set`, but the dotnet CLI uses add/remove/update so we should think in those verbs. 

Note the ordering to match `dotnet add package` and the like commands.

Reply:
Thanks @zivkan for the great suggestion!
I'm thinking if the following sounds good:
`dotnet nuget config list` to list all config values(merged configurations)
`dotnet nuget config list --show-path` to list all config values(merged configurations) and the config file paths.

Reply:
> Apologies for not noticing the ordering and being on autopilot.
> 
> dotnet commands are verb driven.

Thanks @nkolev92 ! I asked Chet about dotnet command conventions and he helped me confirm that: `At this time we have semi-standardized on dotnet NOUN VERB, so dotnet nuget config list would be the preferred form for this command.` (In email: Re: Please advise on naming of a new dotnet nuget command)

Reply:
my bad if I'm providing conflicting feedback :D 

Ignore the ordering then. I'm curious what you think about the other behavior.

Reply:
Thanks @nkolev92 ! Most of them sounds good to me. But seems we still don't have an option to view all config values (what customer mentioned)
So how about the following:

- `dotnet nuget config list`    list all config values (key + value).
- `dotnet nuget config list` --verbosity detailed   list all config values (key + value), and the config file paths.
- `dotnet nuget config list <configKey>`  list the value of the specified config key
- `dotnet nuget config list <configKey> --verbosity detailed`    list the value of the specified config key, and the config file path it comes from
-  `dotnet nuget config add <configKey> <configValue> <config file path>`  add the key value into the specified config file
-  `dotnet nuget config update <configKey> <configValue> <config file path>`  update the key value pair in the specified config file
-  `dotnet nuget config delete <configKey> <config file path>` delete the key value pair from the specified config file

But I'm not sure what shall we do if the configuration is something with child attributes, but not a simple key-value pair. E.g. package source, packageSourceMapping etc. Shall we avoid providing add/update/delete command for those configurations?

Reply:
Should we consider `set` and `unset` instead of `add`, `update`, and `delete`?  That's one less verb, and I don't think that `add` vs `update` matters to customers. (for example, would we want `add` to fail if it already exists in the config?   Would we want to fail `update` if the key doesn't exist in the config file?)

> `dotnet nuget config list <configKey>` list the value of the specified config key

nitpick/very opinionated: A single item is technically a special case of a list, but 

> But I'm not sure what shall we do if the configuration is something with child attributes,

nuget.exe config only handles the [`<configuration><config>` part of the config](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#config-section), which are key-value pairs (probably implemented by `Dictionary<string, string>`). The `nuget.exe config` commands could never be used for other parts of the config file, like sources, credentials, package source mapping, etc.  I'm not saying this is good, but be aware that if you want to make `dotnet nuget config`'s scope so much different to `nuget.exe config`'s existing implementation, it's going to make implementing the spec a lot more difficult.  It's definitely not ideal to design features based on how easy it is to implement, but sometimes we need to be pragmatic. As you pointed out, what would `dotnet nuget config add` commands even look like for package sources, credentials, or package source mapping?

Reply:
> Most of them sounds good to me. But seems we still don't have an option to view all config values (what customer mentioned)

I think we might need a magic word for list all configs such as "all", something we promise to never make a config value for example. 

> Should we consider set and unset instead of add, update, and delete? That's one less verb, and I don't think that add vs update matters to customers. (for example, would we want add to fail if it already exists in the config? Would we want to fail update if the key doesn't exist in the config file?)

I think we should use the same verb as the CLI.

IMO, it's ok with config doesn't handle sources, credentials or source mapping. All of those will have their own individual commands.

Reply:
FWIW, here's a customer complaint that separate add and update verbs are annoying: 
* https://github.com/dotnet/sdk/issues/29026#issue-1445497084

> I think we should use the same verb as the CLI.

Consistency is good, unless it's consistently user unfriendly 😁